pygments/lexers/vyper.py

```diff
@@ -0,0 +1,129 @@
+            (r'\b(bitwise_and|bitwise_not|bitwise_or|bitwise_xor|shift|'
+             r'create_minimal_proxy_to|create_copy_of|create_from_blueprint|'
+             r'ecadd|ecmul|ecrecover|keccak256|sha256|'
+             r'concat|convert|uint2str|extract32|slice|'
+             r'abs|ceil|floor|max|max_value|min|min_value|pow_mod256|sqrt|isqrt|'
+             r'uint256_addmod|uint256_mulmod|unsafe_add|unsafe_sub|unsafe_mul|unsafe_div|'
+             r'as_wei_value|blockhash|empty|len|method_id|_abi_encode|_abi_decode|print|range)\b', Name.Builtin),
+
+            # Built-in Variables and Attributes
+            (r'\b(msg\.sender|msg\.value|block\.timestamp|block\.number|msg\.gas)\b', Name.Builtin.Pseudo),
+
+            # Other variable names and types
+            (r'@internal', Name.Decorator),
+            (r'@[\w.]+', Name.Decorator),
+            (r'__\w+__', Name.Magic),  # Matches double underscores followed by word characters
+            (r'EMPTY_BYTES32', Name.Constant),
+            (r'self\.', Name.Attribute),
+            (r'ERC20', Name.Class),
+            (r'log', Keyword),
+
+            (r'\b(?:u?int\d+|bool|decimal|bytes\d{1,4}|string|String|address|bytes|enum|struct)\b', Keyword.Type),
+
+            (r'\binterface\b', Keyword.Declaration),
+            (r'\bfrom\b', Keyword.Namespace, 'importfrom'),
+
+            # Match function and event signatures
+            (r'\b[a-zA-Z_]\w*\b\([a-zA-Z_]\w*: [a-zA-Z_]\w*\):', Name.Function),
+            (r'\b(?:event|struct)\b \w+', Name.Class, 'body'),
+
+            # Generic names and variables
+            (r'\b[a-zA-Z_]\w*\b:', Name.Variable),
+            (r'\b[a-zA-Z_]\w*\b', Name),
+
+            # Error for unmatched sequences
+            (r'.', Text)
+        ],
+
+        'multiline-comment': [
+            (r'\"\"\"', Comment.Multiline, '#pop'),
+            (r'.|\n', Comment.Multiline)
+        ],
+
+        'importfrom': [
+            (r'\bimport\b', Keyword.Namespace, 'import'),
+            (r'\s+', Text.Whitespace),
+            (r'[a-zA-Z_][\w.]*', Name.Namespace),
+            (r'\s*as\s*', Keyword.Namespace),
+            (r'[a-zA-Z_]\w*', Name.Namespace, '#pop')
+        ],
+
+        'import': [
+            (r'([a-zA-Z_][\w.]*)(\s+as\s+)?([a-zA-Z_]\w*)?',
+            bygroups(Name.Namespace, Keyword.Namespace, Name.Namespace)),
+            (r'\s+', Text.Whitespace),
+            (r',', Punctuation),
+            (r';?', Punctuation, '#pop')
+        ],
+
+        'body': [
+            (r'\(', Punctuation, 'funcparams'),
+            (r':', Punctuation),
+            (r';', Punctuation, '#pop'),
+            (r'\s+', Text.Whitespace),
+            (r'#.*$', Comment.Single)
+        ],
+
+        'funcparams': [
+            (r'\)', Punctuation, '#pop'),
+            (r'\w+', Name.Variable),
+            (r':', Punctuation),
+            (r'\w+', Keyword.Type),
+            (r',', Punctuation),
+            (r'\s+', Text.Whitespace),
```

Comment:
This is the first time I'm seeing someone rely on the fact that the lexer will remember which state it was last in and continue in order.

@birkenfeld , @jeanas : Should we document this better on https://pygments.org/docs/lexerdevelopment/ ? 

It currently says:

> States are groups of regular expressions that are matched against the input string at the current position. If one of these expressions matches, a corresponding action is performed (such as yielding a token with a specific type, or changing state), the current position is set to where the last match ended and the matching process continues with the first regex of the current state.

Which I feel should be: "The process continues using the next rule (in the order of the rule definitions in the lexer source.)" or something, to make this clear. But maybe it's just me given the contributor here figured this out on their own.

Reply:
That's not how it works though? I haven't looked at the source right now but it is intended to work exactly like in the docs, restart at first rule of the state.

If you have a sequence of rules to be matched in order, you should use one rule and bygroups.

Reply:
Ok, then I'm not loosing my mind here. I'll open a PR to clarify this.

Reply:
Also, this needs to be fixed, @mo-anon